nipype/interfaces/base/core.py

```diff
@@ -365,131 +370,42 @@ def run(self, cwd=None, ignore_exception=None, **inputs):
             if successful, results
 
         """
-        from ...utils.profiler import ResourceMonitor
-
-        # if ignore_exception is not provided, taking self.ignore_exception
-        if ignore_exception is None:
-            ignore_exception = self.ignore_exception
-
-        # Tear-up: get current and prev directories
-        syscwd = rgetcwd(error=False)  # Recover when wd does not exist
-        if cwd is None:
-            cwd = syscwd
-
-        os.chdir(cwd)  # Change to the interface wd
-
-        enable_rm = config.resource_monitor and self.resource_monitor
-        self.inputs.trait_set(**inputs)
-        self._check_mandatory_inputs()
-        self._check_version_requirements(self.inputs)
-        interface = self.__class__
-        self._duecredit_cite()
-
-        # initialize provenance tracking
-        store_provenance = str2bool(
-            config.get("execution", "write_provenance", "false")
-        )
-        env = deepcopy(dict(os.environ))
-        if self._redirect_x:
-            env["DISPLAY"] = config.get_display()
-
-        runtime = Bunch(
-            cwd=cwd,
-            prevcwd=syscwd,
-            returncode=None,
-            duration=None,
-            environ=env,
-            startTime=dt.isoformat(dt.utcnow()),
-            endTime=None,
-            platform=platform.platform(),
-            hostname=platform.node(),
-            version=self.version,
+        rtc = RuntimeContext(
+            resource_monitor=config.resource_monitor and self.resource_monitor,
+            ignore_exception=ignore_exception
+            if ignore_exception is not None
+            else self.ignore_exception,
         )
-        runtime_attrs = set(runtime.dictcopy())
-
-        mon_sp = None
-        if enable_rm:
-            mon_freq = float(config.get("execution", "resource_monitor_frequency", 1))
-            proc_pid = os.getpid()
-            iflogger.debug(
-                "Creating a ResourceMonitor on a %s interface, PID=%d.",
-                self.__class__.__name__,
-                proc_pid,
-            )
-            mon_sp = ResourceMonitor(proc_pid, freq=mon_freq)
-            mon_sp.start()
 
-        # Grab inputs now, as they should not change during execution
-        inputs = self.inputs.get_traitsfree()
-        outputs = None
+        with rtc(self, cwd=cwd, redirect_x=self._redirect_x) as runtime:
```

Comment:
This changes the logic of exceptions i believe. previously the run interface only raised/trapped exceptions related to the underlying interface, while other nipype related exceptions would never be skipped by ignore_exception. in the current context, any exception would be either re-raised or ignored.

for example if mandatory checks failed, the interface would raise an exception. now that would be supressed if ignore_exception is true.

this is a change in design which may be worth discussing a bit.

Reply:
I believe we could revert this particular deviation from the original behavior by moving the next three lines outside the context.

I think I added them to make sure they are run while the current directory is `cwd`, but we can use the `indirectory` context Chris wrote some time ago for that.

WDYT?

Reply:
as an initial step, we can do that to maintain current compatibility. although cwd should have no role to play in those three statements, so could simply be run by themselves without `indirectory`.

however, i think the key question here is what are the scenarios, if any, in which your current implementation (which i like) would be troublesome.

Reply:
because contexts behave like try/catch blocks, I think this is pretty safe. I also created the mock resource monitor, so that we don't need to branch depending on whether it is active or not.



> as an initial step, we can do that to maintain current compatibility. although cwd should have no role to play in those three statements, so could simply be run by themselves without `indirectory`.

I'm not sure about this for the ``self.inputs.trait_set(**inputs)`` line. If I'm not wrong, this is going to set those paths in ``inputs``, and if they are provided relative then they will likely fail (unless the user is very aware that the paths should be given relative to the current working directory, not the intended working directory of the interface.

What would you like to see for going forward with this PR?

Reply:
The question is whether there is any reason to ignore the exception if any of the following three lines fail? In the past a failure of those lines would have always generated an exception. Under the current change, those errors can be ignored. 

Also, while unlikely that many people used cwd, for backwards compatibility purposes, we shouldn't adjust the relative path behind the scenes I think.

So to move forward, perhaps simply pull the three statements out of the context manager. That should make the behavior as is.

Reply:
Cool, please note the `os.chdir(cwd)` that used to live a few lines ahead these three... Meaning that we still need to change the directory. Just to make sure we're on the same page.

Reply:
i missed that. in that case `inDirectory` is necessary.

Reply:
Yup

Reply:
(actually that lone chdir was one of the reasons to start this)